core/domain/expression_parser.py

```diff
@@ -0,0 +1,541 @@
+
+    def check_next_token(self):
+        """Function to peek into the next position to see the next token.
+
+        Returns:
+            Token|None. Token at the next position. Returns none if there are no
+                more tokens left.
+        """
+
+        if self.next_token_index < len(self.token_list):
+            return self.token_list[self.next_token_index]
+
+    def get_next_token(self):
+        """Function to retrieve the token at the next position and then
+        increment the next_token_index.
+
+        Returns:
+            Token. Token at the next position.
+
+        Raises:
+            Exception: Invalid syntax: Unexpected end of expression.
+        """
+
+        if self.next_token_index < len(self.token_list):
+            token = self.token_list[self.next_token_index]
+            self.next_token_index += 1
+            return token
+
+        raise Exception('Invalid syntax: Unexpected end of expression.')
+
+    def is_next_token(self, allowed_token_texts):
+        """Function to verify that there is at least one more token remaining
+        and that the next token is among the allowed_token_texts provided.
+
+        Args:
+            allowed_token_texts: list(str). List of strings containing the
+                allowed token texts at the next position.
+
+        Returns:
+            Token|None. Token at the next position. Returns none if there are no
+                more tokens left or the next token is not in the
+                allowed_token_texts.
+        """
+
+        if self.next_token_index < len(self.token_list):
+            text = self.token_list[self.next_token_index].text
+            if text in allowed_token_texts:
+                token = self.token_list[self.next_token_index]
+                self.next_token_index += 1
+                return token
+        return None
+
+    def expect_token(self, text):
+        """Function to compare the next token with the given text. If
+        comparision is true, increments next_token_index.
+
+        Args:
+            text: str. String representation of the token to check.
+
+        Returns:
+            Token. Token at the next position.
+
+        Raises:
+            Exception: Invalid token.
+        """
+        token = self.check_next_token()
+        if token is None or token.text != text:
+            raise Exception('Invalid token: %s.' % (
+                'None' if token is None else token.text))
+
+        token = self.token_list[self.next_token_index]
+        self.next_token_index += 1
+        return token
+
+
+def is_valid_expression(expression):
+    """Checks if given math expression is syntactically valid.
+
+    Args:
+        expression: str. String representation of the math expression.
+
+    Returns:
+        bool. Whether the given math expression is syntactically valid.
+    """
+
+    try:
+        parser = Parser(expression)
+        parser.parse()
+        return True
```

Comment:
Have return True outside try - except. That would be the same, right?

Reply:
Yup, done.